File: Labs/Lab09 - 10 std- teacher- course/src/main/java/com/efekurucay/lab09/services/StudentService.java
package com.efekurucay.lab09.services;

import com.efekurucay.lab09.model.Course;
import com.efekurucay.lab09.model.Student;
import com.efekurucay.lab09.repository.CourseRepository;
import com.efekurucay.lab09.repository.StudentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class StudentService {
    
    private final StudentRepository studentRepository;
    private final CourseRepository courseRepository;
    
    @Autowired
    public StudentService(StudentRepository studentRepository, CourseRepository courseRepository) {
        this.studentRepository = studentRepository;
        this.courseRepository = courseRepository;
    }
    
    // Tüm öğrencileri getir
    public List<Student> getAllStudents() {
        return studentRepository.findAll();
    }
    
    // ID'ye göre öğrenci getir
    public Student getStudentById(Long id) {
        return studentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Öğrenci bulunamadı: " + id));
    }
    
    // Öğrenci numarasına göre öğrenci getir
    public Student getStudentByNumber(String studentNumber) {
        return studentRepository.findByStudentNumber(studentNumber)
                .orElseThrow(() -> new RuntimeException("Öğrenci numarası ile öğrenci bulunamadı: " + studentNumber));
    }
    
    // Yeni öğrenci oluştur
    public Student createStudent(Student student) {
        return studentRepository.save(student);
    }
    
    // Öğrenciyi güncelle
    public Student updateStudent(Long id, Student studentDetails) {
        Student student = getStudentById(id);
        student.setName(studentDetails.getName());
        student.setSurname(studentDetails.getSurname());
        student.setEmail(studentDetails.getEmail());
        student.setStudentNumber(studentDetails.getStudentNumber());
        student.setTeacher(studentDetails.getTeacher());
        return studentRepository.save(student);
    }
    
    // Öğrenciyi sil
    public void deleteStudent(Long id) {
        studentRepository.deleteById(id);
    }
    
    // Bir öğrencinin tüm derslerini getir
    public List<Course> getStudentCourses(Long studentId) {
        Student student = getStudentById(studentId);
        return student.getCourses();
    }
    
    // Öğrencinin bekleyen derslerini getir
    public List<Course> getPendingCourses(Long studentId) {
        Student student = getStudentById(studentId);
        return student.getCourses().stream()
                .filter(course -> Course.CourseStatus.PENDING.equals(course.getStatus()))
                .collect(Collectors.toList());
    }
    
    // Öğrencinin onaylanmış derslerini getir
    public List<Course> getApprovedCourses(Long studentId) {
        Student student = getStudentById(studentId);
        return student.getCourses().stream()
                .filter(course -> Course.CourseStatus.APPROVED.equals(course.getStatus()))
                .collect(Collectors.toList());
    }
    
    // Öğrenciye ders ekle (default olarak PENDING durumunda)
    public Course addCourseToStudent(Long studentId, Course course) {
        Student student = getStudentById(studentId);
        course.setStatus(Course.CourseStatus.PENDING);
        Course savedCourse = courseRepository.save(course);
        
        List<Course> courses = student.getCourses();
        courses.add(savedCourse);
        student.setCourses(courses);
        studentRepository.save(student);
        
        return savedCourse;
    }
    
    // Mevcut bir derse kayıt ol
    public Course enrollCourse(Long studentId, Long courseId) {
        Student student = getStudentById(studentId);
        Course course = courseRepository.findById(courseId)
                .orElseThrow(() -> new RuntimeException("Ders bulunamadı: " + courseId));
        
        // Öğrencinin zaten bu derse kayıtlı olup olmadığını kontrol et
        if (student.getCourses().contains(course)) {
            throw new RuntimeException("Öğrenci zaten bu derse kayıtlı: " + course.getName());
        }
        
        // Dersi PENDING durumunda ekle
        course.setStatus(Course.CourseStatus.PENDING);
        
        // Öğrenci-ders ilişkisini güncelle
        List<Course> courses = student.getCourses();
        courses.add(course);
        student.setCourses(courses);
        studentRepository.save(student);
        
        return course;
    }
}

File: Labs/Lab09 - 10 std- teacher- course/frontend/src/app/services/student.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Course } from '../models/course';
import { Student } from '../models/student';

@Injectable({
  providedIn: 'root'
})
export class StudentService {
  private apiUrl = environment.apiUrl + '/api/student';

  constructor(private http: HttpClient) { }

  // Tüm öğrencileri getir
  getAllStudents(): Observable<Student[]> {
    return this.http.get<Student[]>(this.apiUrl);
  }

  // ID'ye göre öğrenci getir
  getStudentById(id: number): Observable<Student> {
    return this.http.get<Student>(`${this.apiUrl}/${id}`);
  }

  // Öğrenci oluştur
  createStudent(student: Student): Observable<Student> {
    return this.http.post<Student>(this.apiUrl, student);
  }

  // Öğrenci güncelle
  updateStudent(id: number, student: Student): Observable<Student> {
    return this.http.put<Student>(`${this.apiUrl}/${id}`, student);
  }

  // Öğrenci sil
  deleteStudent(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Öğrencinin tüm derslerini getir
  getStudentCourses(studentId: number): Observable<Course[]> {
    return this.http.get<Course[]>(`${this.apiUrl}/courses/all?studentId=${studentId}`);
  }

  // Öğrencinin bekleyen derslerini getir
  getPendingCourses(studentId: number): Observable<Course[]> {
    return this.http.get<Course[]>(`${this.apiUrl}/courses/pending?studentId=${studentId}`);
  }

  // Öğrencinin onaylanmış derslerini getir
  getApprovedCourses(studentId: number): Observable<Course[]> {
    return this.http.get<Course[]>(`${this.apiUrl}/courses/approved?studentId=${studentId}`);
  }

  // Tüm mevcut dersleri getir (öğrencinin kaydolabileceği dersler)
  getAllAvailableCourses(): Observable<Course[]> {
    return this.http.get<Course[]>(`${this.apiUrl}/courses/available`);
  }

  // Öğrenci için yeni ders ekle
  addCourse(studentId: number, course: Course): Observable<Course> {
    return this.http.post<Course>(`${this.apiUrl}/courses/add?studentId=${studentId}`, course);
  }
  
  // Mevcut bir derse kayıt ol
  enrollCourse(studentId: number, courseId: number): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/courses/enroll`, {
      studentId: studentId,
      courseId: courseId
    });
  }
} 